packages/framework-provider-azure/src/library/events-adapter.ts

```diff
@@ -72,35 +78,69 @@ export async function readEntityLatestSnapshot(
       snapshot
     )
-    return snapshot as EventEnvelope
+    return snapshot as EntitySnapshotEnvelope
   } else {
     logger.debug(
       `[EventsAdapter#readEntityLatestSnapshot] No snapshot found for entity ${entityTypeName} with ID ${entityID}.`
     )
-    return null
+    return undefined
   }
 }
 
 export async function storeEvents(
   cosmosDb: CosmosClient,
-  eventEnvelopes: Array<EventEnvelope>,
+  eventEnvelopes: Array<NonPersistedEventEnvelope>,
   config: BoosterConfig
 ): Promise<void> {
   const logger = getLogger(config, 'events-adapter#storeEvents')
   logger.debug('[EventsAdapter#storeEvents] Storing EventEnvelopes with eventEnvelopes:', eventEnvelopes)
   for (const eventEnvelope of eventEnvelopes) {
+    const persistableEvent: EventEnvelope = {
+      ...eventEnvelope,
+      persistedAt: new Date().toISOString(),
+    }
     await cosmosDb
       .database(config.resourceNames.applicationStack)
       .container(config.resourceNames.eventsStore)
       .items.create({
-        ...eventEnvelope,
+        ...persistableEvent,
         [eventsStoreAttributes.partitionKey]: partitionKeyForEvent(
           eventEnvelope.entityTypeName,
-          eventEnvelope.entityID,
-          eventEnvelope.kind
+          eventEnvelope.entityID
         ),
-        [eventsStoreAttributes.sortKey]: new Date().toISOString(),
+        [eventsStoreAttributes.sortKey]: persistableEvent.persistedAt,
       })
   }
   logger.debug('[EventsAdapter#storeEvents] EventEnvelope stored')
 }
+
+export async function storeSnapshot(
+  cosmosDb: CosmosClient,
+  snapshotEnvelope: EntitySnapshotEnvelope,
+  config: BoosterConfig
+): Promise<void> {
+  const logger = getLogger(config, 'events-adapter#storeSnapshot')
+  logger.debug('[EventsAdapter#storeSnapshot] Storing snapshot with snapshotEnvelope:', snapshotEnvelope)
+
+  const partitionKey = partitionKeyForSnapshot(snapshotEnvelope.entityTypeName, snapshotEnvelope.entityID)
+  /**
+   * The sort key of the snapshot matches the sort key of the last event that generated it.
+   * Entity snapshots can be potentially created by competing processes, so we need to make sure
+   * that no matter what snapshot we find, we can always rebuild the correct state by
+   * replaying all the events that happened after the snapshot we take as the origin.
+   * This way of storing the data makes snapshot caching an idempotent operation, and allows us to
+   * aggressively caching snapshots without worrying about insertion order.
+   */
+  const sortKey = snapshotEnvelope.snapshottedEventPersistedAt
+
+  await cosmosDb
+    .database(config.resourceNames.applicationStack)
+    .container(config.resourceNames.eventsStore)
+    .items.create({
+      ...snapshotEnvelope,
+      [eventsStoreAttributes.partitionKey]: partitionKey,
+      [eventsStoreAttributes.sortKey]: sortKey,
+    })
+
+  logger.debug('[EventsAdapter#storeSnapshot] Snapshot stored')
```

Comment:
Could we improve this log?

```suggestion
  logger.debug('Snapshot stored', snapshotEnvelope)
```